kernel/sched.c

```diff
@@ -397,15 +399,43 @@ static void move_thread_to_end_of_prio_q(struct k_thread *thread)
 	update_cache(thread == _current);
 }
 
-static void flag_ipi(void)
+static void flag_ipi(uint32_t ipi_mask)
 {
 #if defined(CONFIG_SMP) && defined(CONFIG_SCHED_IPI_SUPPORTED)
 	if (arch_num_cpus() > 1) {
-		_kernel.pending_ipi = true;
+		atomic_or(&_kernel.pending_ipi, (atomic_val_t)ipi_mask);
 	}
 #endif
 }
 
+/* Create a bitmask of CPUs that need an IPI. Note: sched_spinlock is held. */
+static atomic_val_t ipi_mask_create(struct k_thread *thread)
+{
+#if defined(CONFIG_SMP) && (CONFIG_MP_MAX_NUM_CPUS > 1)
+	uint32_t  ipi_mask = 0;
+	struct k_thread *cpu_thread;
+	unsigned int num_cpus = arch_num_cpus();
+	uint32_t  id = arch_proc_id();
+
+	for (uint32_t i = 0; i < num_cpus; i++) {
+		if (id == i) {
+			continue;
+		}
+
+		cpu_thread = _kernel.cpus[i].current;
+		if ((cpu_thread != NULL) &&
+		    (cpu_thread->base.prio > thread->base.prio)) {
```

Comment:
Nice catch! I had not considered deadlines. Using z_sched_prio_cmp() may be a better choice over a straight priority comparison, but I will have to mull this over some more with an eye for races with deadlines (such musings at this time are night are prone to error).

Ignoring the inclusion of deadlines for the moment (only because I haven't given that enough thought yet and it is about 11 PM where I am), I am quite sure that this does not introduce a new scenario where something could happen in the unlocked interim that could confuse things and require an IPI to be delivered. Aside: this does not fix the issue described in #69005, but I don't think that it makes it worse, and I was planning on fixing that one in a separate PR.

Still, I get the concern about stuff going wrong here--it can be dang hard to get it right, and very easy to miss something. Assuming that this is not dead in the water, I fully expect that there will be some back and forth discussions as the ideas gets explored and hopefully it stands up to attempts to rip it to shreds--it is important to get it right.

Reply:
Now that it is day time and I managed to get _some_ sleep last night, I have been taking (yet) another look and I need to take a closer look at the following two routines (and who calls them)--z_set_prio() and k_thread_deadline_set().

The routine z_set_prio() is called by z_thread_priority_set(), and that is not looking like it is quite right for SMP--which in turn would affect the behaviour here. As for k_thread_deadline_set(), it seems to me that changing the deadline such that it results in a lower effective priority should result in a reschedule on potentially more than 1 core as there may now be a thread in the ready queue that  should take precedence.